src/txn/txn.c

```diff
@@ -1620,17 +1626,42 @@ __wt_txn_commit(WT_SESSION_IMPL *session, const char *cfg[])
         if (cursor != NULL)
             WT_CLEAR(cursor->key);
     }
-    txn->mod_count = 0;
-#ifdef HAVE_DIAGNOSTIC
-    WT_ASSERT(session, txn->prepare_count == prepare_count);
-    txn->prepare_count = 0;
-#endif
 
     if (cursor != NULL) {
         WT_ERR(cursor->close(cursor));
         cursor = NULL;
     }
 
+    /*
+     * Resolve any fast-truncate transactions and allow eviction to proceed on instantiated pages.
+     * This isn't done as part of the initial processing because until now the commit could still
+     * switch to an abort. The action allowing eviction to proceed is clearing the WT_UPDATE list,
+     * (if any), associated with the commit. We're the only consumer of that list and we no longer
+     * need it, and eviction knows it means abort or commit has completed on instantiated pages.
+     */
+    for (i = 0, op = txn->mod; ft_resolution > 0 && i < txn->mod_count; i++, op++)
+        if (op->type == WT_TXN_OP_REF_DELETE) {
+            __wt_txn_op_set_timestamp(session, op);
+
+            WT_REF_LOCK(session, op->u.ref, &previous_state);
+            if (previous_state == WT_REF_DELETED)
+                op->u.ref->ft_info.del->committed = 1;
+            else
+                __wt_free(session, op->u.ref->ft_info.update);
+            WT_REF_UNLOCK(op->u.ref, previous_state);
+
+            __wt_txn_op_free(session, op);
+
+            --ft_resolution;
+        }
+    WT_ASSERT(session, ft_resolution == 0);
```

Comment:
ft_resolution is just being used as a sanity check?

Reply:
Yes and no; the assert is just a sanity check, but the counter `ft_resolution` avoids processing some number of slots, that is, once we're done, we can avoid processing any more slots. We could tighten it down even more (start at the first slot we have to process, so skip leading slots we don't care about),  but I didn't bother.

Reply:
Right, yeah, should've noticed the early loop exit. From my recollection the txn mod's are sorted so a start slot would definitely avoid walking a longer list of entries.

Reply:
Ah its sorted by btree-id and then key, so not mod type. So even with a start slot you'd still potentially be walking a long way. Maybe not a lot of point adding it then.